sophiread/SophireadCLI/src/sophiread_core.cpp

```diff
@@ -0,0 +1,384 @@
+    const std::vector<std::vector<std::vector<TIFF32Bit>>>& tof_images,
+    const std::vector<double>& tof_bin_edges,
+    const std::string& tof_filename_base)
+{
+    auto start = std::chrono::high_resolution_clock::now();
+
+    // 1. Create output directory if it doesn't exist
+    if (!std::filesystem::exists(out_tof_imaging)) {
+        std::filesystem::create_directories(out_tof_imaging);
+        spdlog::info("Created output directory: {}", out_tof_imaging);
+    }
+
+    // 2. Initialize vector for spectral data
+    std::vector<uint64_t> spectral_counts(tof_images.size(), 0);
+
+    // 3. Iterate through each TOF bin and save TIFF files
+    tbb::parallel_for(tbb::blocked_range<size_t>(0, tof_images.size()), [&](const tbb::blocked_range<size_t>& range) {
+      for (size_t bin = range.begin(); bin < range.end(); ++bin) {
+          // Construct filename
+          std::string filename = fmt::format("{}/{}_bin_{:04d}.tiff", out_tof_imaging, tof_filename_base, bin + 1);
+
+          // prepare container and fill with current hist2d
+          const uint32_t width = tof_images[bin][0].size();
+          const uint32_t height = tof_images[bin].size();
+          std::vector<std::vector<TIFF32Bit>> accumulated_image = tof_images[bin];
+
+          // check if file already exist
+          if (std::filesystem::exists(filename)) {
+              TIFF* existing_tif = TIFFOpen(filename.c_str(), "r");
+              if (existing_tif) {
+                  uint32_t existing_width, existing_height;
+                  TIFFGetField(existing_tif, TIFFTAG_IMAGEWIDTH, &existing_width);
+                  TIFFGetField(existing_tif, TIFFTAG_IMAGELENGTH, &existing_height);
+
+                  if (existing_width == width && existing_height == height) {
+                      // Dimensions match, proceed with accumulation
+                      for (uint32_t row = 0; row < height; ++row) {
+                          std::vector<TIFF32Bit> scanline(width);
+                          TIFFReadScanline(existing_tif, scanline.data(), row);
+                          for (uint32_t col = 0; col < width; ++col) {
+                              accumulated_image[row][col] += scanline[col];
+                          }
+                      }
+                      spdlog::debug("Accumulated counts for existing file: {}", filename);
+                  } else {
+                      spdlog::error("Dimension mismatch for file: {}. Expected {}x{}, got {}x{}. Overwriting.",
+                                    filename, width, height, existing_width, existing_height);
+                  }
+                  TIFFClose(existing_tif);
+              } else {
+                  spdlog::error("Failed to open existing TIFF file for reading: {}", filename);
+              }
+          }
+
+          // Write or update TIFF file
+          TIFF* tif = TIFFOpen(filename.c_str(), "w");
+          if (tif) {
+              uint32_t width = tof_images[bin][0].size();
```

Comment:
If this part is slow, mark these two variables `const` and they'll get optimized out

Reply:
thanks, I added the `const` to both var